src/runner/PlutoRunner.jl

```diff
@@ -50,33 +57,156 @@ function set_current_module(newname)
-const table_column_display_limit = 8
-const table_column_display_limit_increase = 30
+const computer_workspace = Main
 
-const tree_display_extra_items = Dict{UUID, Dict{ObjectDimPair, Int64}}()
+function register_computer(expr::Expr, input_globals::Vector{Symbol}, output_globals::Vector{Symbol})
+    proof = ReturnProof()
 
-function formatted_result_of(id::UUID, ends_with_semicolon::Bool, showmore::Union{ObjectDimPair,Nothing}=nothing)::NamedTuple{(:output_formatted, :errored, :interrupted, :runtime),Tuple{MimedOutput,Bool,Bool,Union{UInt64, Missing}}}
-    extra_items = if showmore === nothing
-        tree_display_extra_items[id] = Dict{ObjectDimPair, Int64}()
+    @gensym result
+    e = Expr(:function, Expr(:tuple, input_globals...), Expr(:block, 
+        Expr(:(=), result, timed_expr(expr, proof)),
+        Expr(:tuple,
+            result,
+            Expr(:tuple, output_globals...)
+        )
+    ))
+
+    f = Core.eval(computer_workspace, e)
+
+    computers[expr] = Computer(f, proof, input_globals, output_globals)
+end
+
+function compute(computer::Computer)
+    input_global_values = getfield.([current_module], computer.input_globals)
+
+    out = Base.invokelatest(computer.f, input_global_values...)
+    if out isa Tuple{Any,Tuple}
+        result, output_global_values = out
+
+        for (name, val) in zip(computer.output_globals, output_global_values)
+            Core.eval(current_module, Expr(:(=), name, val))
+        end
+
+        result
     else
-        old = get!(() -> Dict{ObjectDimPair, Int64}(), tree_display_extra_items, id)
-        old[showmore] = get(old, showmore, 0) + 1
-        old
+        throw(return_error)
     end
+end
 
-    ans = cell_results[id]
-    errored = ans isa CapturedException
-    output_formatted = (!ends_with_semicolon || errored) ? format_output(ans; context=:extra_items=>extra_items) : ("", MIME"text/plain"())
-    (output_formatted = output_formatted, errored = errored, interrupted = false, runtime = Main.runtime)
+function run_expression(expr::Any, cell_id::UUID, function_wrapped_info::Union{Nothing,Tuple{Set{Symbol},Set{Symbol}}}=nothing)
+    cell_results[cell_id], cell_runtimes[cell_id] = if function_wrapped_info === nothing
+        trycatch_expr(expr, cell_id)
+    else
+        computer = get!(computers, expr) do
+            register_computer(expr, collect.(function_wrapped_info)...)
+        end
+        trycatch_expr(cell_id, computer.return_proof) do
+            compute(computer)
+        end
+    end
+end
+
+
+"Wrap `expr` inside a timing block."
+function timed_expr(expr::Expr, return_proof::Any=nothing)::Expr
+    # @assert ExpressionExplorer.is_toplevel_expr(expr)
+
+    linenumbernode = expr.args[1]
+    root = expr.args[2] # pretty much equal to what `Meta.parse(cell.code)` would give
+
+    @gensym result
+    @gensym elapsed_ns
+    # we don't use `quote ... end` here to avoid the LineNumberNodes that it adds (these would taint the stack trace).
```

Comment:
you could use `MacroTools.striplines(ex)` for removing LineNumberNotes from expressions generated by `quote` (if this would be beneficial for you).